### PR TITLE
Add bitwise shift operations

### PR DESCRIPTION
**Pull request by gastonar**
Allows bitwise shift operations including arithmetic shift right, logical shift right, and shift left.

### CODE REVIEW

**Comment by jessesleeping** [comment]
This is a duplicate PR (original one: #12886). I suggest we close this one and use the original one as we already had review comments there.

**Comment by tdcmeehan** [comment]
> This is a duplicate PR (original one: #12886). I suggest we close this one and use the original one as we already had review comments there.

I recommended that we do that because there was a tricky rebase issue.  The other one is closed and we can address that comment here.

**Comment by tdcmeehan** [comment]
Other functions in `MathFunctions` also implement other overloads for numeric types with smaller ranges.  This is because the engine will automatically widen smaller types into larger ones.  For example, without an overload, the engine will happily expand the function into `BIGINT` if the underlying data is `TINYINT`.  I think we can add some quick overloads for smaller numeric types so this doesn't happen.

Now, all of these functions will take in Java type of long, which means after shifting it's possible that the value exceeds the range of the type.  But luckily we already specify the bits we want shifted, so we could adjust the max range to match the type.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestBitwiseFunctions.java [hunk 1]
```diff
@@ -90,4 +92,44 @@ public void testBitwiseXor()
...
+
+    @Test
+    public void testBitwiseSrl()
+    {
+        assertFunction("bitwise_shift_right_logical(7, 2, 4)", BIGINT, 1L);
+        assertFunction("bitwise_shift_right_logical(7, 2, 64)", BIGINT, 7L >>> 2L);
+        assertFunction("bitwise_shift_right_logical(-4, 6, 64)", BIGINT, -4L >>> 6L);
+        assertFunction("bitwise_shift_right_logical(-8, 2, 5)", BIGINT, 6L);
+        //-1152905011916701696 = 0xF0000F0000F00000
+        assertFunction("bitwise_shift_right_logical(-1152905011916701696, 62, 64)", BIGINT, 3L);
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestBitwiseFunctions.java]:
```suggestion
        assertFunction("bitwise_shift_right_logical(0xF0000F0000F00000L, 62, 64)", BIGINT, 3L);
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/BitwiseFunctions.java [hunk 2]
```diff
@@ -75,4 +75,60 @@ public static long bitwiseXor(@SqlType(StandardTypes.BIGINT) long left, @SqlType
...
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Bits specified in bit_count must be between 2 and 64, got " + bits);
+        }
+
+        if (shift < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Specified shift must be positive");
+        }
+
+        number = number << (64 - bits);
+
+        return number >>> (64 - bits + shift) & (long) (Math.pow(2, bits) - 1);
```
[the remark this answers is not in the document]
Comment (jessesleeping) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/BitwiseFunctions.java]: Or can we just do a bitwise mask and then run the logical shift right? e.g.
```java
        return (number & (long) (Math.pow(2, bits) - 1)) >>> shift;
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/BitwiseFunctions.java [hunk 3]
```diff
@@ -75,4 +75,60 @@ public static long bitwiseXor(@SqlType(StandardTypes.BIGINT) long left, @SqlType
...
+    @SqlType(StandardTypes.BIGINT)
+    public static long bitwiseShiftLeftLogical(@SqlType(StandardTypes.BIGINT) long number, @SqlType(StandardTypes.BIGINT) long shift,
+            @SqlType(StandardTypes.BIGINT) long bits)
+    {
+        if (bits == 64) {
+            return number << shift;
+        }
+
+        if (bits <= 1 || bits > 64) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Bits specified in bit_count must be between 2 and 64, got " + bits);
```
[the remark this answers is not in the document]
Comment (gastonar) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/BitwiseFunctions.java]: In previous bitwise functions, specifically bit_count, the input wasn't allowed to have a single bit, so I kept it like that for consistency. I can change it to simply being `bits < 1` though.

File: presto-docs/src/main/sphinx/functions/bitwise.rst [hunk 4]
```diff
@@ -28,4 +28,27 @@ Bitwise Functions
 
     Returns the bitwise XOR of ``x`` and ``y`` in 2's complement representation.
 
+.. function:: bitwise_shift_left_logical(x, shift, bits) -> bigint
```
Comment (rongrong) [inline presto-docs/src/main/sphinx/functions/bitwise.rst]: I personally found the `bits` parameter really confusing. If `x` is a `smallint`, am I supposed to know how many bits a `smallint` should have and provide the right value for `bits` to get a correct result?

**Comment by rongrong** [comment]
Personally I'd like the names to be `logical_shift_left`, `logical_shift_right`, `arithmetic_shift_left`, `arithmetic_shift_right`. "bitwise" should be implicit since it's a shift operation. I don't think it's absolutely necessary, but we can add it to the beginning if people feel that's better.

File: presto-docs/src/main/sphinx/functions/bitwise.rst [hunk 5]
```diff
@@ -28,4 +28,27 @@ Bitwise Functions
...
+
+    Shift left logical operation on ``x`` (treated as ``bits``-bit signed integer)
+    shifted by ``shift`` in 2's complement representation.
+
+        SELECT bitwise_shift_left_logical(7, 2, 4); -- 12
```
Comment (rongrong) [inline presto-docs/src/main/sphinx/functions/bitwise.rst]: 12 is a 5-bit signed number. This doesn't make sense to me. if 12 is a 4-bit signed number, then we cannot represent a 64-bit signed number with bigint.

Comment (rongrong) [reply presto-docs/src/main/sphinx/functions/bitwise.rst]: (not necessarily that operation doesn't make sense, maybe just the description doesn't make sense).

Comment (rongrong) [reply presto-docs/src/main/sphinx/functions/bitwise.rst]: Overall, for logical left shift, you probably will always fill 0, so `bits` is how many bits to keep after shift (where to truncate). So describing the number as n-bit signed just adds confusion.

For logical right shift, what you want to describe is what to fill with (0 or 1), so maybe defining the sign-bit makes sense. Though I think it makes more sense to say how many bits I want to operate on, and whether I'd fill it with 0 or 1 (`logical_right_shift(number, shift, bits, 0/1)`) Describing logical shift with signs is just very confusing to me.

**Comment by rongrong** [comment]
Related to #4028

File: presto-docs/src/main/sphinx/functions/bitwise.rst (hunk 5 above)
Comment (rongrong) [reply presto-docs/src/main/sphinx/functions/bitwise.rst]: If right shift is also going to fill with 0, then we should not describe the function using "signed number" at all.

**Comment by tdcmeehan** [comment]
> Personally I'd like the names to be `logical_shift_left`, `logical_shift_right`, `arithmetic_shift_left`, `arithmetic_shift_right`. "bitwise" should be implicit since it's a shift operation. I don't think it's absolutely necessary, but we can add it to the beginning if people feel that's better.

I think they should include `bitwise` in the front to be consistent with the other bitwise functions.

https://prestodb.github.io/docs/current/functions/bitwise.html

There would be no difference between arithmetic and logic left shift.  So we're left with:

* `bitwise_shift_left`
* `bitwise_arithmetic_right_shift`
* `bitwise_logical_right_shift`

i.e. logical/arithmetic is now in front rather than back of the name

File: presto-docs/src/main/sphinx/functions/bitwise.rst (hunk 4 above)
Comment (tdcmeehan) [reply presto-docs/src/main/sphinx/functions/bitwise.rst]: The bits convention is already a established with `bit_count`: https://prestodb.github.io/docs/current/functions/bitwise.html

I think, while confusing, it should be consistent; or, we deprecate the `bits` convention.  But for now I'd stick with consistency.

File: presto-docs/src/main/sphinx/functions/bitwise.rst (hunk 5 above)
Comment (gastonar) [reply presto-docs/src/main/sphinx/functions/bitwise.rst]: By definition, logical right shift will always fill in 0s, so it does not preserve the sign of the input. Meanwhile, arithmetic right shift will fill with the MSB of the input. This operation will preserve the sign of the input. Therefore, I think we should just stick to these default conventions. I feel like adding the `0/1` field as a 4th input would just make it more confusing.

Comment (gastonar) [reply presto-docs/src/main/sphinx/functions/bitwise.rst]: I did remove the "signed number" part in the description though, I can see how that can be confusing.

**Comment by rongrong** [comment]
I'm wondering, why are we going with functions rather than operator? Ideally, there should be `>>` and `<<` and depending on the type of the input (signed or unsigned), it would be logical or arithmetic. Though it doesn't seem that Presto support unsigned type, which makes it a bit inconvenient. But arguably, in a system that does not support unsigned numbers, it's strange to support logical shift.

**Comment by highker** [comment]
I guess it's because SQL syntax only has basic numeric operators specified (+, -, ...). For those not in SQL spec, it could be safer to treat them as functions for possible future compliance. 
```
valueExpression
    : primaryExpression                                                                 #valueExpressionDefault
    | valueExpression AT timeZoneSpecifier                                              #atTimeZone
    | operator=(MINUS | PLUS) valueExpression                                           #arithmeticUnary
    | left=valueExpression operator=(ASTERISK | SLASH | PERCENT) right=valueExpression  #arithmeticBinary
    | left=valueExpression operator=(PLUS | MINUS) right=valueExpression                #arithmeticBinary
    | left=valueExpression CONCAT right=valueExpression                                 #concatenation
    ;

**Comment by rongrong** [comment]
Dug around a little bit and Postgres support `>>` and `<<` and Oracle actually has `>>`, `<<` and `>>>`. So I'm betting on SQL Spec not assigning conflicting meanings to these in the foreseeable future. That can be a separate issue though. With that we can potentially still have a use of these functions, since they allow arbitrary `bits` rather than just valid data type boundaries. Operator, on the other hand, would return you the correct type, rather than always make it a `BIGINT` and require down-casting. In summary, I think it's worth exploring the operator route, but I won't block this PR from being merged.

**Pull request merged**